File: scala/src/test/scala/PortfolioShould.scala
import domain.{Bank, Money, Portfolio}
import org.scalatest.EitherValues

class PortfolioShould
    extends org.scalatest.funsuite.AnyFunSuite
    with EitherValues {
  private val bank = Bank()
    .addExchangeRate("EUR", "USD", 1.2)
    .addExchangeRate("USD", "KRW", 1100)

  test("5 USD + 10 USD = 15 USD") {
    val fiveDollars = Money(5, "USD")
    val tenDollars = Money(10, "USD")
    val portfolio = Portfolio(fiveDollars, tenDollars)

    assert(portfolio.evaluate(bank, "USD").value === Money(15, "USD"))
  }

  test("5 USD + 10 EUR = 17 USD") {
    val fiveDollars = Money(5, "USD")
    val tenEuros = Money(10, "EUR")
    val portfolio = Portfolio(fiveDollars, tenEuros)

    assert(portfolio.evaluate(bank, "USD").value === Money(17, "USD"))
  }

  test("1 USD + 1100 KRW = 2200 KRW") {
    val portfolio = Portfolio(Money(1, "USD"), Money(1100, "KRW"))
    assert(portfolio.evaluate(bank, "KRW").value === Money(2200, "KRW"))
  }

  test("Return a greedy message in case of missing exchange rates") {
    val portfolio = Portfolio(Money(1, "USD"), Money(1, "EUR"), Money(1, "KRW"))
    assert(
      portfolio
        .evaluate(bank, "KRW")
        .left
        .value === "Missing exchange rate(s): [EUR->KRW]"
    )
  }
}

File: c#/money-problem.Tests/PortfolioShould.cs
using System;
using FluentAssertions;
using money_problem.Domain;
using Xunit;
using static money_problem.Domain.Types.Currency;

namespace money_problem.Tests
{
    public class PortfolioShould
    {
        private readonly Types.Bank _bank =
            Bank.WithExchangeRate(EUR, USD, 1.2)
                .AddExchangeRate(USD, KRW, 1100);

        [Fact(DisplayName = "5 USD + 10 USD = 15 USD")]
        public void Add()
        {
            var portfolio = 5d.Dollars().AddToPortfolio(10d.Dollars());
            portfolio.Evaluate(_bank, USD)
                .RightUnsafe()
                .Should()
                .Be(15d.Dollars());
        }

        [Fact(DisplayName = "5 USD + 10 EUR = 17 USD")]
        public void AddDollarsAndEuros()
        {
            var portfolio = 5d.Dollars().AddToPortfolio(10d.Euros());
            portfolio.Evaluate(_bank, USD)
                .RightUnsafe()
                .Should()
                .Be(17d.Dollars());
        }

        [Fact(DisplayName = "1 USD + 1100 KRW = 2200 KRW")]
        public void AddDollarsAndKoreanWons()
        {
            var portfolio = 1d.Dollars().AddToPortfolio(1100d.KoreanWons());
            portfolio.Evaluate(_bank, KRW)
                .RightUnsafe()
                .Should()
                .Be(2200d.KoreanWons());
        }

        [Fact(DisplayName = "Return a Left in case of missing exchange rates")]
        public void AddWithMissingExchangeRatesShouldReturnALeft()
        {
            var portfolio = 1d.Dollars()
                .AddToPortfolio(1d.Euros())
                .Add(1d.KoreanWons());

            portfolio.Evaluate(_bank, KRW)
                .LeftUnsafe()
                .Should()
                .Be("Missing exchange rate(s): [EUR->KRW]");
        }
    }
}